compiler/src/iree/compiler/Utils/Permutation.h

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_PERMUTATION_H_
+#define IREE_COMPILER_UTILS_PERMUTATION_H_
+
+#include <iterator>
+#include <utility>
+
+#include "llvm/ADT/ADL.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Support/LLVM.h"
+
+namespace mlir::iree_compiler {
+
+// Example: values = (1, 2, 3), permutation = (2, 1, 0)
+// output = (3, 2, 1).
+template <typename ValuesIt, typename PermutationRange, typename OutIt>
+void permute(ValuesIt valuesBegin, ValuesIt valuesEnd,
```

Comment:
There are some permutation related helpers in https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h. Wanna double check whether we can use some of them?

Reply:
The function `applyPermutation` in MLIR is not generic and accepts only int64_t for the permutation value type. I put a generic one in `Utils/Permutation.h` with a comment to make the one upstream generic.